andela/firestar-backend
# #167730744 Implements Reset password feature

#### What does this PR do?
* Resets user password by sending a cryptographically stamped token to the email of a registered user on request. The token allows a user to set new password. Only a valid user can reset passwords, invalid requests are sent links to sign up on the website
#### Description of Task to be completed?
* None
#### How should this be manually tested?
* git clone this repo https://github.com/andela/firestar-backend.git
* cd to `firestar-backend`
* git checkout `ft-reset-password-167730744`
* run `npm install`

* set PORT on .env to 4000
* set TOKENEXPIRY to any integer value (describes the expiration time for your reset password link in seconds)
* CLIENT_URL links to an heroku url where the api is served or any local server
* DB_USERNAME=
*DB_PASSWORD=
*DB_NAME=
*DB_TEST_NAME=
*NODE_ENV=test
 and
> Register for an account on sendgrid.com and create API key at https://app.sendgrid.com/settings/api_keys, then set the following environment variables
>* SENDGRID_API_KEY

Also in your .env, supply this field with valid information to receive reset link in your email
YOUR_EMAIL="" (description: Creates a user account with this email to receive reset password link)
SOME_PASSWORD="" (description: Create user account with password, Must include more than 3 characters)

* DO ONLY npm test (to run tests)

OR
* npm run migrate:dev (to create necessary database tables) 
* npm run seed:dev (to populate a test database )
* run npm start:dev (to start development server)

Then startup POSTMAN and
* go to localhost:4000/api/v1/users/passwords/forgot (HTTP METHOD: POST) and supply one of the emails you sync with (in your YOUR_EMAIL env)
* then you should get an email with a rest link
* then copy the link in your email into postman, set {password and confirmPassword}
* also try invalid data to catch errors

To test resetpassword
* From the link in your email, copy the link to postman and run a POST request

#### Any background context you want to provide?
* Users may be lose access to their account, but are able to retrieve it with their email. Emails provide proof of authentication and obfuscate the membership of the network to only the members.

#### What are the relevant pivotal tracker stories?
* [#167730744](https://www.pivotaltracker.com/n/projects/2382004/stories/167730744)

#### Screenshots or gifs (Code coverage)
![image](https://user-images.githubusercontent.com/28393660/64321173-f05b7080-cfb7-11e9-878a-a12ecd18a893.png)

![image](https://user-images.githubusercontent.com/28393660/63473027-f6a40580-c46b-11e9-9f58-b59ef90bc228.png)

![image](https://user-images.githubusercontent.com/28393660/64132506-4b2e7580-cdc8-11e9-8932-e9b2dc0f2462.png)

#### Questions:
* N/A
